### PR TITLE
 staticanalysis/bot: Allow justifying clang-tidy checkers with a 'reason' text field (#1481)

### PR DESCRIPTION
**Pull request by rv404674**
Added a text field, to justify use of checkers in reveiwbot comments.

### CODE REVIEW

File: src/staticanalysis/bot/tests/mocks/config.yaml [hunk 1]
```diff
@@ -7,6 +7,7 @@ clang_checkers:
    publish: !!bool yes
  - name: clang-analyzer-security.*
    publish: !!bool no
+   reason: "Mandated by Mozilla's style guide: https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style#Control_Structures"
```
Comment (jankeromnes) [inline src/staticanalysis/bot/tests/mocks/config.yaml]: Nit: Even though it would work for a test file, I find it a bit strange to say that "all security checkers are mandated by Mozilla's style guide".

Could you please move the reason field below, e.g. to "modernize-use-nullptr", and change its text to something like "Modernize our code base to C++11"? That would make more sense to me.

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 2]
```diff
@@ -235,6 +235,12 @@ def __init__(self, header_data, revision):
         assert len(header_data) == 6
         assert not settings.repo_dir.endswith('/')
         self.revision = revision
+
+        self.reason = None
+        for c in settings.clang_checkers:
+            if c['name'] == self.check:
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: As you can see below your code, `self.check` is only set to the right value (from `header_data`) *after* this code runs, so it looks like unfortunately your code reads "find all the checkers where `name` is `None`".

To fix this, please move your entire code block below, e.g. at the end of the `__init__` method.

**Comment by rv404674** [comment]
hey @jankeromnes , it is completed. Can you please tell me one thing, you have said to check is self.reason is None. That means i have to check in all of the  export methods, i.e. __str__, as_text, as_markdown, as_diff, and as_dic

**Comment by jankeromnes** [comment]
> hey @jankeromnes , it is completed.

Awesome, thanks! I'll try to have a look today.

> Can you please tell me one thing, you have said to check is self.reason is None. That means i have to check in all of the export methods, i.e. str, as_text, as_markdown, as_diff, and as_dic

Yes, I think that in all of these methods, you'll have to do something like:

```python
if self.reason is not None:
    output += "\nReason: {}".format(self.reason)
```

(Or maybe you can simply do `if self.reason: ...` ? I'm a bit worried about what happens if `self.reason` is not `None` but is the empty string, e.g. because of an unforeseen bug somewhere.)

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 3]
```diff
@@ -297,11 +302,17 @@ def as_text(self):
         Build the text body published on reporters
         '''
         message = self.message
+        reason = self.reason
+
         if len(message) > 0:
             message = message[0].capitalize() + message[1:]
-        body = '{}: {} [clang-tidy: {}]'.format(
+        if len(reason) > 0:
+            reason = reason[0].capitalize + reason[1:]
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: I don't think you need to capitalize `reason` (which we edit ourselves in `config.yaml`) the same way as `message` (which is generated by `clang-tidy`).

Please don't do that, we'll try to remember to always begin our `reason` texts with a capital letter.

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 4]
```diff
@@ -297,11 +302,17 @@ def as_text(self):
...
+
+        body = '{}: {} {} [clang-tidy: {}]'.format(
             self.type.capitalize(),
             message,
+            reason,
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: I don't think you want to cram the `reason` field into the first line of the comment. Instead, please add the reason after `self.body` (see below), with a similar conditional check (i.e. `if self.reason: body += '\n{}'.format(self.reason)`).

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 5]
```diff
@@ -244,8 +244,13 @@ def __init__(self, header_data, revision):
...
         self.notes = []
 
+        self.reason = None
+        for c in settings.clang_checkers:
+            if c['name'] == self.check:
+                self.reason = c['reason']
+
     def __str__(self):
-        return '[{}] {} {} {}:{}'.format(self.type, self.check, self.path, self.line, self.char)
+        return '[{}] {} {} {} {}:{}'.format(self.type, self.check, self.reason, self.path, self.line, self.char)
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: Since we don't include `self.body` here, let's also not include `self.reason`, in order to keep the output short.

**Comment by rv404674** [comment]
hey @jankeromnes , Done. :)

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 6]
```diff
@@ -308,13 +315,16 @@ def as_text(self):
...
+        if self.reason:
+            body += '\n{}'.format(self.reason)
 
         return body
 
     def as_markdown(self):
         return ISSUE_MARKDOWN.format(
             type=self.type,
             message=self.message,
+            reason=self.reason,
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: Please move `reason` just below the `body`. I feel it would make more sense there.

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 7]
```diff
@@ -351,6 +361,7 @@ def as_dict(self):
             'check': self.check,
             'type': self.type,
             'message': self.message,
+            'reason': self.reason,
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: Please move `'reason'` just below `'body'`.

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 8]
```diff
@@ -244,6 +244,11 @@ def __init__(self, header_data, revision):
         self.body = None
         self.notes = []
 
+        self.reason = None
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: I think `None` is the default value for unknown attributes. No need to initialize it here.

**Comment by rv404674** [comment]
hey @jankeromnes , how to run all these check locally - i mean for build failure, indention checks.. etc.
I use '.please check staticanalysis/bot' but it takes lot of time.

**Comment by jankeromnes** [comment]
@La0 or @garbas, do you know if we have a faster way to run checks on a project than `./please check`? E.g. maybe just the `flake8` linting?

**Comment by rv404674** [comment]
hey @janx, little bit help in find what is problem in tests. Sorry for bothering you again and again. Thanks

**Comment by La0** [comment]
In the nix shell, you can simply use `flake8` to validate linting, and `pytest` to run all the unit tests.

Looks like your code broke some unit tests about missing `reason` in some `Issue` classes implementations.

**Comment by rv404674** [comment]
hey can you come in IRC, will be more easy to talk there. :)

**Comment by La0** [comment]
@rv404674 I'm on IRC, but working on different projects right now.

**Comment by jankeromnes** [comment]
@rv404674 You can just look in This pull request > `All checks have failed` > `Details` > `[failed] 1.02. Building staticanalysis/bot` > `Run Logs` > `.../live.log` > (scroll to bottom).

This currently shows the following errors:

```
    def as_text(self):
        '''
            Build the text body published on reporters
            '''
        message = self.message
        if len(message) > 0:
            message = message[0].capitalize() + message[1:]
        body = '{}: {} [clang-tidy: {}]'.format(
            self.type.capitalize(),
            message,
            self.check,
        )

        # Always add body as it's been cleaned up
        if self.body:
            body += '\n{}'.format(self.body)
>       if self.reason:
E       AttributeError: 'ClangTidyIssue' object has no attribute 'reason'
```

and

```
    def as_markdown(self):
        return ISSUE_MARKDOWN.format(
            type=self.type,
            message=self.message,
            location='{}:{}:{}'.format(self.path, self.line, self.char),
            body=self.body,
>           reason=self.reason,
            check=self.check,
            in_patch='yes' if self.revision.contains(self) else 'no',
            third_party='yes' if self.is_third_party() else 'no',
            publishable_check='yes' if self.has_publishable_check() else 'no',
            publishable='yes' if self.is_publishable() else 'no',
            expanded_macro='yes' if self.is_expanded_macro() else 'no',
            is_new='yes' if self.is_new else 'no',
            notes='\n'.join([
                ISSUE_NOTE_MARKDOWN.format(
                    message=n.message,
                    location='{}:{}:{}'.format(n.path, n.line, n.char),
                    body=n.body,
                ) for n in self.notes
            ]),
        )
E       AttributeError: 'ClangTidyIssue' object has no attribute 'reason'
```

Maybe you need to do something like `if self.reason is not None:`?

Or maybe I was wrong to tell you that setting `self.reason = None` wasn't necessary in `__init__` (in which case, sorry for that!)

**Comment by rv404674** [comment]
hey @jankeromnes , instead of pushing to upstream all the times. Now i will first check it locally, 
using these in order.
'./please shell staticanalysis/bot'
then 
'pytest'
and 'flake 8' , no???. and thus i will bother you less also. :)

**Comment by rv404674** [comment]
hey @jankeromnes, setting 'self.reason = None' , worked. You can finally merge it. :)

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 9]
```diff
@@ -244,6 +244,11 @@ def __init__(self, header_data, revision):
...
+        for c in settings.clang_checkers:
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: This part will not work in every case, you need to support wildcard matches (for example we use `mozilla-*` to describe all checks which name start with `mozilla-`.

Please look [this code](https://github.com/mozilla/release-services/blob/4625854c25aa6630184751a85325d30de2756c8d/src/staticanalysis/bot/static_analysis_bot/config.py#L106), and add a generic function such that `settings.get_check(check_name)` returns a check. You can then use it here and in the `is_publishable_check` function.

**Comment by rv404674** [comment]
hey @La0 , can you please take a look at my new commit and help me in how to resolve the error.

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 10]
```diff
@@ -307,7 +315,9 @@ def as_text(self):
 
         # Always add body as it's been cleaned up
         if self.body:
-            body += '\n```\n{}\n```'.format(self.body)
+            body += '\n{}'.format(self.body)
```
Comment (jankeromnes) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: Nit: Here, you're accidentally reverting a change that I did recently. That's why the tests fail (if you look at the failed test logs, you'll see that it complains about missing "```" in the output).

**Comment by jankeromnes** [comment]
Also, please don't forget to address the last paragraph of https://github.com/mozilla/release-services/pull/1556#pullrequestreview-159728405

**Comment by rv404674** [comment]
hey @jankeromnes , can you please take a look,when you get some time. Thanks. :)

File: src/staticanalysis/bot/static_analysis_bot/clang/tidy.py [hunk 11]
```diff
@@ -244,6 +244,14 @@ def __init__(self, header_data, revision):
         self.body = None
         self.notes = []
 
+        self.reason = None
+        for c in settings.clang_checkers:
+            name = c['name']
+
+            # Wildcard at the end of check name and also normal matching
+            if name == self.check or (name.endswith('*') and self.check.startswith(name[:-1])):
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/clang/tidy.py]: You duplicated the code from [settings.is_publishable_check](https://github.com/mozilla/release-services/blob/4625854c25aa6630184751a85325d30de2756c8d/src/staticanalysis/bot/static_analysis_bot/config.py#L97) instead of creating a common function shared between `settings.is_publishable_check` and here, as requested.

Please do the following steps:

1. create a common function in [settings.py](https://github.com/mozilla/release-services/blob/4625854c25aa6630184751a85325d30de2756c8d/src/staticanalysis/bot/static_analysis_bot/config.py#L97), named `get_clang_check`
2. `get_clang_check` must be the only place where this decision code is present
3. `get_clang_check` must return the dict containing all information of the required check description, or `None` if no check is present
4. Update `settings.is_publishable_check` to use `get_clang_check`. Its behaviour must remain the same.
5. Finally use `get_clang_check` here like so:

```python
check = settings.get_clang_check(self.check)
if check is not None:
  self.reason = check.get('reason')
```

File: src/staticanalysis/bot/tests/test_clang.py [hunk 12]
```diff
@@ -148,9 +148,11 @@ def test_clang_tidy(mock_repository, mock_config, mock_clang, mock_stats, mock_r
     assert len(issues) == 2
     assert isinstance(issues[0], ClangTidyIssue)
     assert issues[0].check == 'modernize-use-nullptr'
+    assert issues[0].reason == 'Modernize our code base to C++11'
     assert issues[0].line == 3
     assert isinstance(issues[1], ClangTidyIssue)
     assert issues[1].check == 'modernize-use-nullptr'
+    assert issues[1].reason == 'Modernize our code base to C++11'
```
Comment (La0) [inline src/staticanalysis/bot/tests/test_clang.py]: Bonus point if you add another case in the unit test to test a check *without* a reason (and check that `reason is None`)

File: src/staticanalysis/bot/static_analysis_bot/config.py [hunk 13]
```diff
@@ -103,18 +103,32 @@ def is_publishable_check(self, check):
...
+            return None
+
+        reason_dict = {}
         for c in self.clang_checkers:
             name = c['name']
 
             if name.endswith('*') and check.startswith(name[:-1]):
                 # Wildcard at end of check name
-                return c.get('publish', True)
+                reason_dict['reason'] = c['reason']
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/config.py]: `return c` directly here, so we get the full data

File: src/staticanalysis/bot/static_analysis_bot/config.py [hunk 14]
```diff
@@ -103,18 +103,32 @@ def is_publishable_check(self, check):
...
 
             elif name == check:
                 # Same exact check name
-                return c.get('publish', True)
+                reason_dict['reason'] = c['reason']
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/config.py]: `return c` directly here, so we get the full data

File: src/staticanalysis/bot/static_analysis_bot/config.py [hunk 15]
```diff
@@ -103,18 +103,32 @@ def is_publishable_check(self, check):
...
 
-        return False
+        return reason_dict
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/config.py]: `return None` by default

File: src/staticanalysis/bot/static_analysis_bot/config.py [hunk 16]
```diff
@@ -103,18 +103,32 @@ def is_publishable_check(self, check):
         if check is None:
             return False
 
+        temp_dict = self.is_publishable_check(check)
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/config.py]: This code should look like the following, so it behaves as before

```python
clang_check = self.get_clang_check(check)
return clang_check.get('publish', True) if clang_check else False
```

File: src/staticanalysis/bot/tests/test_clang.py (hunk 12 above)
Comment (rv404674) [reply src/staticanalysis/bot/tests/test_clang.py]: Hey @La0 with the test. How to trigger a clang tidy error in using BAD_SRC_CPP. Can you take a look at irc Please.

**Comment by rv404674** [comment]
yippeeeee, hey @La0 , it is finally done.  The bug was simply wrong function name. Please do take a look when you get some free time.
